sdk/core/Azure.Core.Experimental/api/Azure.Core.Experimental.net461.cs

```diff
@@ -111,97 +111,152 @@ namespace Azure.Core.Dynamic
-            public Azure.Core.Dynamic.DynamicJson.ObjectEnumerator GetEnumerator() { throw null; }
+            public Azure.Core.Dynamic.DynamicData.ObjectEnumerator GetEnumerator() { throw null; }
             public bool MoveNext() { throw null; }
             public void Reset() { }
-            System.Collections.Generic.IEnumerator<Azure.Core.Dynamic.DynamicJsonProperty> System.Collections.Generic.IEnumerable<Azure.Core.Dynamic.DynamicJsonProperty>.GetEnumerator() { throw null; }
+            System.Collections.Generic.IEnumerator<Azure.Core.Dynamic.DynamicDataProperty> System.Collections.Generic.IEnumerable<Azure.Core.Dynamic.DynamicDataProperty>.GetEnumerator() { throw null; }
             System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() { throw null; }
         }
     }
-    public enum DynamicJsonNameMapping
+    public enum DynamicDataNameMapping
     {
         None = 0,
         PascalCaseGetters = 1,
         PascalCaseGettersCamelCaseSetters = 2,
     }
     [System.Runtime.InteropServices.StructLayoutAttribute(System.Runtime.InteropServices.LayoutKind.Sequential)]
-    public partial struct DynamicJsonOptions
+    public partial struct DynamicDataOptions
     {
         private int _dummyPrimitive;
-        public static readonly Azure.Core.Dynamic.DynamicJsonOptions AzureDefault;
-        public DynamicJsonOptions() { throw null; }
-        public Azure.Core.Dynamic.DynamicJsonNameMapping PropertyNameCasing { get { throw null; } set { } }
+        public static readonly Azure.Core.Dynamic.DynamicDataOptions AzureDefault;
+        public DynamicDataOptions() { throw null; }
+        public Azure.Core.Dynamic.DynamicDataNameMapping PropertyNameCasing { get { throw null; } set { } }
     }
     [System.Runtime.InteropServices.StructLayoutAttribute(System.Runtime.InteropServices.LayoutKind.Sequential)]
-    public readonly partial struct DynamicJsonProperty
+    public readonly partial struct DynamicDataProperty
     {
         private readonly object _dummy;
         private readonly int _dummyPrimitive;
         public string Name { get { throw null; } }
-        public Azure.Core.Dynamic.DynamicJson Value { get { throw null; } }
+        public Azure.Core.Dynamic.DynamicData Value { get { throw null; } }
+    }
+    public abstract partial class ObjectDocument : System.IDisposable
+    {
+        protected ObjectDocument() { }
+        public abstract T As<T>(object element);
+        public abstract void Dispose();
+        public abstract Azure.Core.Dynamic.ObjectElement GetIndexElement(object element, int index);
+        public abstract bool HasValue(object element);
+        public abstract void Set(object element, object value);
+        public abstract Azure.Core.Dynamic.ObjectElement SetProperty(object element, string name, object value);
+        public abstract string? ToString(object element);
+        public abstract bool TryGetArrayLength(object element, out int length);
```

Comment:
Instead of ValueKind, whether or not the element is a given type is achieved through TryGet methods.  These will be on ObjectElement as well, with more variety.

Reply:
It might be the best approach, but there is a tradeoff. In my deserialization code, I will have to keep trying many Try calls and possibly only the last one will succeed.

Reply:
I see, yes that makes sense.  I can add a ValueKind enum instead - we will need to decide what those are ahead of time.

Reply:
One observation here is that we end up back at the "JSON has fewer types than .NET" issue.  If we introduce an `ObjectValueKind` enum, we could map values to the smallest subset of types (e.g. JSON's number instead of int, long, double, etc) and let callers TryGet() to determine which .NET types are valid.  We could use CLR types, and adopt a convention that we assign kind the largest valid type and assume callers are able to downcast if they need it to be a different type.

Reply:
Or we could do something like:

```csharp
    public enum ObjectValueKind
    {
        Undefined,
        Object,
        Array,
        String,
        Integer,
        FloatingPoint,
        Boolean,
        Null
    }
```

Where we split JSON's Number into [Integer](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/integral-numeric-types) and [FloatingPoint](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/floating-point-numeric-types) and let users downcast from those.

Reply:
Updated to reflect this.  I'm not sure how this works in the case where you have 1-1 fidelity with CLR types (would you want to know it's an int and not a long in the calling code?), or where the set of types you're mapping to is potentially more expressive.